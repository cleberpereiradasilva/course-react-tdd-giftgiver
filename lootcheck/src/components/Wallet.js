import React, { Component } from 'react'
import { connect } from 'react-redux'
import { deposit, withdraw } from '../actions/balance'


import { 
    Form, 
    FormGroup, 
    FormControl,     
    Col,
    Row,
    Button 
} from 'react-bootstrap'

export class Wallet extends Component{
    constructor(){
        super()
        this.state = {
            balance: undefined
        }
    }
    
    updateWallet = event => {
        this.setState({
            balance: parseInt(event.target.value, 10)
        })
    }
    deposit = () => this.props.deposit(this.state.balance)
    withdraw = () => this.props.withdraw(this.state.balance)
    
    render() {
        return (            
                <Row className='gift'>
                    <Form className='gift__form'>
                        <div className='row'>
                            <Col md={12}>
                                <h3 className='span-balance'>Wallet balance: {this.props.balance}</h3>
                            </Col>
                            <Col md={12}>
                            <FormGroup>                                
                                <FormControl className='input-wallet'                         
                                    onChange={this.updateWallet}  />
                            </FormGroup>
                            </Col>
                            <Col md={6} className='gift__form__row__button'>
                                <Button type='button' className='btn btn-success btn-deposit' 
                                onClick={this.deposit} >
                                    +
                                </Button>
                            </Col>
                            <Col md={6} className='gift__form__row__button'>
                                <Button type='button' className='btn btn-danger btn-withdraw' 
                                onClick={this.withdraw} >
                                    -
                                </Button>
                            </Col>
                            
                        </div>

                    </Form>
            </Row>
        )
    }
}

export default connect(state => { return { balance: state.balance }}, { deposit, withdraw })(Wallet)
