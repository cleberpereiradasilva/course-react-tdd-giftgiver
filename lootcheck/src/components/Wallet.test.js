import React from 'react'
import { shallow } from 'enzyme'
import  { Wallet } from './Wallet'

describe('Wallet', () => {
    const mockDeposit = jest.fn()
    const mockWithdraw = jest.fn()
    const props = { balance: 20, deposit: mockDeposit, withdraw: mockWithdraw }
    const wallet = shallow(<Wallet {...props}/>)
    
    it('check if `Wallet` exist', () => {
        expect(wallet).toMatchSnapshot()
        
    })

    it('check if `balance` is visible', () => {        
        expect(wallet.find('.span-balance').text()).toEqual(`Wallet balance: ${ props.balance }`)
    })

    it('create input to deposit into or withdraw from balance', () => {        
        expect(wallet.find('.input-wallet').exists()).toBe(true)
    })
    
    describe('when the user type into input', () => {
        const userBalance = '25'
        beforeEach(() => {
            wallet.find('.input-wallet').simulate('change', {
                target: { value: userBalance }
            })            
        })
        it('update local wallet balance', () => {                               
            expect(wallet.state().balance).toEqual(parseInt(userBalance, 10))
        })
        describe('and user wants to make a deposit', () => {
            beforeEach(() => wallet.find('.btn-deposit').simulate('click'))
            it('dispatches `deposit`', () => {        
                expect(mockDeposit).toHaveBeenCalledWith(parseInt(userBalance,10))
            })
        })


        describe('and user wants to witdhdraw', () => {
            beforeEach(() => wallet.find('.btn-withdraw').simulate('click'))
            it('dispatches `withdraw`', () => {        
                expect(mockWithdraw).toHaveBeenCalledWith(parseInt(userBalance,10))
            })
        })



    })
})
