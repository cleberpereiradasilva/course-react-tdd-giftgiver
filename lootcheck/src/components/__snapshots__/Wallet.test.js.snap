// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wallet check if \`Wallet\` exist 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <Row
    bsClass="row"
    className="gift"
    componentClass="div"
>
    <Form
        bsClass="form"
        className="gift__form"
        componentClass="form"
        horizontal={false}
        inline={false}
    >
        <div
            className="row"
        >
            <Col
                bsClass="col"
                componentClass="div"
                md={12}
            >
                <h3
                    className="span-balance"
                >
                    Wallet balance: 
                    20
                </h3>
            </Col>
            <Col
                bsClass="col"
                componentClass="div"
                md={12}
            >
                <FormGroup
                    bsClass="form-group"
                >
                    <ControlLabel
                        bsClass="control-label"
                        srOnly={false}
                    >
                        Present
                    </ControlLabel>
                    <FormControl
                        bsClass="form-control"
                        className="input-wallet"
                        componentClass="input"
                        onChange={[Function]}
                    />
                </FormGroup>
            </Col>
            <Col
                bsClass="col"
                className="gift__form__row__button"
                componentClass="div"
                md={6}
            >
                <Button
                    active={false}
                    block={false}
                    bsClass="btn"
                    bsStyle="default"
                    className="btn btn-danger btn-deposit"
                    disabled={false}
                    onClick={[Function]}
                    type="button"
                >
                    +
                </Button>
            </Col>
            <Col
                bsClass="col"
                className="gift__form__row__button"
                componentClass="div"
                md={6}
            >
                <Button
                    active={false}
                    block={false}
                    bsClass="btn"
                    bsStyle="default"
                    className="btn btn-danger btn-withdraw"
                    disabled={false}
                    onClick={[Function]}
                    type="button"
                >
                    -
                </Button>
            </Col>
        </div>
    </Form>
</Row>,
  "nodes": Array [
    <Row
      bsClass="row"
      className="gift"
      componentClass="div"
>
      <Form
            bsClass="form"
            className="gift__form"
            componentClass="form"
            horizontal={false}
            inline={false}
      >
            <div
                  className="row"
            >
                  <Col
                        bsClass="col"
                        componentClass="div"
                        md={12}
                  >
                        <h3
                              className="span-balance"
                        >
                              Wallet balance: 
                              20
                        </h3>
                  </Col>
                  <Col
                        bsClass="col"
                        componentClass="div"
                        md={12}
                  >
                        <FormGroup
                              bsClass="form-group"
                        >
                              <ControlLabel
                                    bsClass="control-label"
                                    srOnly={false}
                              >
                                    Present
                              </ControlLabel>
                              <FormControl
                                    bsClass="form-control"
                                    className="input-wallet"
                                    componentClass="input"
                                    onChange={[Function]}
                              />
                        </FormGroup>
                  </Col>
                  <Col
                        bsClass="col"
                        className="gift__form__row__button"
                        componentClass="div"
                        md={6}
                  >
                        <Button
                              active={false}
                              block={false}
                              bsClass="btn"
                              bsStyle="default"
                              className="btn btn-danger btn-deposit"
                              disabled={false}
                              onClick={[Function]}
                              type="button"
                        >
                              +
                        </Button>
                  </Col>
                  <Col
                        bsClass="col"
                        className="gift__form__row__button"
                        componentClass="div"
                        md={6}
                  >
                        <Button
                              active={false}
                              block={false}
                              bsClass="btn"
                              bsStyle="default"
                              className="btn btn-danger btn-withdraw"
                              disabled={false}
                              onClick={[Function]}
                              type="button"
                        >
                              -
                        </Button>
                  </Col>
            </div>
      </Form>
</Row>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <Wallet
        balance={20}
        deposit={[Function]}
        withdraw={[Function]}
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": Wallet {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "deposit": [Function],
        "props": Object {
          "balance": 20,
          "deposit": [Function],
          "withdraw": [Function],
        },
        "refs": Object {},
        "state": Object {
          "balance": undefined,
        },
        "updateWallet": [Function],
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
        "withdraw": [Function],
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <Row
          bsClass="row"
          className="gift"
          componentClass="div"
>
          <Form
                    bsClass="form"
                    className="gift__form"
                    componentClass="form"
                    horizontal={false}
                    inline={false}
          >
                    <div
                              className="row"
                    >
                              <Col
                                        bsClass="col"
                                        componentClass="div"
                                        md={12}
                              >
                                        <h3
                                                  className="span-balance"
                                        >
                                                  Wallet balance: 
                                                  20
                                        </h3>
                              </Col>
                              <Col
                                        bsClass="col"
                                        componentClass="div"
                                        md={12}
                              >
                                        <FormGroup
                                                  bsClass="form-group"
                                        >
                                                  <ControlLabel
                                                            bsClass="control-label"
                                                            srOnly={false}
                                                  >
                                                            Present
                                                  </ControlLabel>
                                                  <FormControl
                                                            bsClass="form-control"
                                                            className="input-wallet"
                                                            componentClass="input"
                                                            onChange={[Function]}
                                                  />
                                        </FormGroup>
                              </Col>
                              <Col
                                        bsClass="col"
                                        className="gift__form__row__button"
                                        componentClass="div"
                                        md={6}
                              >
                                        <Button
                                                  active={false}
                                                  block={false}
                                                  bsClass="btn"
                                                  bsStyle="default"
                                                  className="btn btn-danger btn-deposit"
                                                  disabled={false}
                                                  onClick={[Function]}
                                                  type="button"
                                        >
                                                  +
                                        </Button>
                              </Col>
                              <Col
                                        bsClass="col"
                                        className="gift__form__row__button"
                                        componentClass="div"
                                        md={6}
                              >
                                        <Button
                                                  active={false}
                                                  block={false}
                                                  bsClass="btn"
                                                  bsStyle="default"
                                                  className="btn btn-danger btn-withdraw"
                                                  disabled={false}
                                                  onClick={[Function]}
                                                  type="button"
                                        >
                                                  -
                                        </Button>
                              </Col>
                    </div>
          </Form>
</Row>,
        "_debugID": 2,
        "_renderedOutput": <Row
          bsClass="row"
          className="gift"
          componentClass="div"
>
          <Form
                    bsClass="form"
                    className="gift__form"
                    componentClass="form"
                    horizontal={false}
                    inline={false}
          >
                    <div
                              className="row"
                    >
                              <Col
                                        bsClass="col"
                                        componentClass="div"
                                        md={12}
                              >
                                        <h3
                                                  className="span-balance"
                                        >
                                                  Wallet balance: 
                                                  20
                                        </h3>
                              </Col>
                              <Col
                                        bsClass="col"
                                        componentClass="div"
                                        md={12}
                              >
                                        <FormGroup
                                                  bsClass="form-group"
                                        >
                                                  <ControlLabel
                                                            bsClass="control-label"
                                                            srOnly={false}
                                                  >
                                                            Present
                                                  </ControlLabel>
                                                  <FormControl
                                                            bsClass="form-control"
                                                            className="input-wallet"
                                                            componentClass="input"
                                                            onChange={[Function]}
                                                  />
                                        </FormGroup>
                              </Col>
                              <Col
                                        bsClass="col"
                                        className="gift__form__row__button"
                                        componentClass="div"
                                        md={6}
                              >
                                        <Button
                                                  active={false}
                                                  block={false}
                                                  bsClass="btn"
                                                  bsStyle="default"
                                                  className="btn btn-danger btn-deposit"
                                                  disabled={false}
                                                  onClick={[Function]}
                                                  type="button"
                                        >
                                                  +
                                        </Button>
                              </Col>
                              <Col
                                        bsClass="col"
                                        className="gift__form__row__button"
                                        componentClass="div"
                                        md={6}
                              >
                                        <Button
                                                  active={false}
                                                  block={false}
                                                  bsClass="btn"
                                                  bsStyle="default"
                                                  className="btn btn-danger btn-withdraw"
                                                  disabled={false}
                                                  onClick={[Function]}
                                                  type="button"
                                        >
                                                  -
                                        </Button>
                              </Col>
                    </div>
          </Form>
</Row>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Wallet
    balance={20}
    deposit={[Function]}
    withdraw={[Function]}
/>,
}
`;
